markovflow/kalman_filter.py

```diff
@@ -495,3 +494,119 @@ def _log_det_observation_precision(self):
     def observations(self):
         """ Observation vector """
         return self.sites.means
+
+
+@tf_scope_class_decorator
+class KalmanFilterWithSparseSites(BaseKalmanFilter):
+    r"""
+    Performs a Kalman filter on a :class:`~markovflow.state_space_model.StateSpaceModel`
+    and :class:`~markovflow.emission_model.EmissionModel`, with Gaussian sites, over a time grid.
+    """
+
+    def __init__(self, state_space_model: StateSpaceModel, emission_model: EmissionModel, sites: GaussianSites,
+                 time_grid_shape: tf.TensorShape, observations_index: tf.Tensor, observations: tf.Tensor):
+        """
+        :param state_space_model: Parameterises the latent chain.
+        :param emission_model: Maps the latent chain to the observations.
+        :param sites: Gaussian sites over the time grid.
+        :param time_grid_shape: Time grid shape, tuple (T, 1).
+        :param observations_index: Index of the observations in the time grid with shape (N,).
+        :param observations: Observations with shape (N, state_dim).
+        """
+        self.sites = sites
+        self.time_grid_shape = time_grid_shape
+        self.observations_index = observations_index
+        self._observations = tf.scatter_nd(self.observations_index, observations, self.time_grid_shape)
+        self.data_sites = self._get_data_sites()
+        super().__init__(state_space_model, emission_model)
```

Comment:
```suggestion
        self.sparse_observations = observations
        self.data_sites = self._get_data_sites()
        super().__init__(state_space_model, emission_model)
        
    @property
    def _observations(self):
        """
        The sparse vector of observation of size on the time grid.
        """
        return tf.scatter_nd(self.observations_index, self.sparse_observations, self.time_grid_shape)

```

Reply:
I think it is best to save the sparse vector and to have a method to expand it

then you don't need the gather_nd later on

Reply:
Currently, how it is done is that we calculate `disp = observations - marginals` and then gather on these `disp` so saving sparse_observations won't to much as then we will need to do gather on marginals anyways.

Reply:
I see what you mean here now.

Reply:
Incorporated it this way, the benefit is that we only save sparse observations but for 'disp' and `disp_data` we do sparse_to_dense and vice versa.